Improve WooLog entry list

0nko opened the pull request:
This PR improves the `WooLog` rolling entry list by using a more efficient data structure, namely ~~`LinkedHashMap`~~`LinkedList`. 

The `WooLog` limits the number of entries to a certain number of elements (99). Previously the last entry was manually removed. ~~With the new approach, this is done automattically using the `LinkedHashMap.removeEldestEntry()` function.~~

The main reason for this PR is to avoid using an `ArrayList` as a base class to implement `LogEntryList` because it's causing a crash due to a bug in Jetpack Compose:

```
 No virtual method getSize()I in class Lcom/woocommerce/android/util/WooLog$LogEntryList; or its super classes (declaration of ‘com.woocommerce.android.util.WooLog$LogEntryList’)
```

0nko committed:
Use LinkedHashMap to implement the rolling entry list

(cherry picked from commit 96a1a2dbcb4a9ec694c8a9cb911d4da1fa0e61ae)

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/WooLog.kt:
```diff
@@ -44,7 +43,7 @@ object WooLog {
     private const val MAX_ENTRIES = 99
     private val logEntries = LogEntryList()
-    private val listeners = ArrayList<LogListener>(0)
+    private val listeners = mutableListOf<LogListener>()
```
It looks like that we actually never use those listeners and the code is since 2018. So maybe to remove the code related to this?

0nko replied:
Yeah, I noticed that but I thought it'd be safer to keep, it just in case. But I'm fine with getting rid of it. 👍

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/WooLog.kt:
```diff
@@ -217,60 +216,43 @@ object WooLog {
-            }
-            return sb.toString()
-        }
+        override fun toString() = values.toList().joinToString("\n")
```
Nice improvement 👍

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/WooLog.kt:
```diff
@@ -217,60 +216,43 @@ object WooLog {
-            }
+        @Synchronized
+        fun add(element: LogEntry) {
+            put(size, element)
```
I would add a unit test on this because TBH it's not really clear how it works for me. So at some moment size will be always == 99 and you will be rewriting the entry with key = 99? It feels like we should use some incremental id here instead of a "size"

0nko replied:
Good call! While working on the unit tests I realized the `LinkedHashMap` didn't work as expected and in the end, using a `LinkedList` works without having to mess with a hash key. Most of all, it still works with Jetpack Compose.

kidinov reviewed (commented):
@0nko Nice improvement! It should work somewhat faster than an array

 Please check my comments, I think the one about the  usage of size as a key might be worth applying a change

I had the same issue with Compose. Do you have any clue why this might happen?

0nko committed:
Remove unused code

0nko committed:
Move LogEntryList to a separate class and use a list instead of a map

0nko committed:
Remove unused code

0nko committed:
Use the native kotlin println()

0nko committed:
Add the missing import

0nko committed:
Rename the log list class and throw an exception if the limit is <= 0

0nko committed:
Add a unit test for log entries

0nko committed:
Fix detekt issues

0nko commented:
Thanks for the review @kidinov! I've added some unit tests and changed the `LinkedHashMap` to `LinkedList` because it didn't work properly.

> I had the same issue with Compose. Do you have any clue why this might happen?

I'm not sure. I didn't find any information about it 🤷.

Ready for another round.

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/util/RollingLogEntriesTest.kt:
```diff
@@ -0,0 +1,56 @@
+
+class RollingLogEntriesTest {
+    @Test
+    fun `Verify all entries are present if limit not reached`() {
```
nit: Several months ago we managed to agree on test [naming convention ](https://github.com/woocommerce/woocommerce-android/blob/develop/CONVENTION.md#unit-tests) which might be good to follow. It's pretty much:

`given, when, then` pattern with commas, e.g.:

```
fun `given the button has 9 clicks, when the button is clicked, then the appropriate message is shown`() {
```

0nko replied:
Oh yeah, I forgot about that.. Thanks!

kidinov reviewed (approved):
Looks great! 👍 

I just left a comment regarding the naming of the test which is optional. Please go ahead with merging with that change or without

0nko committed:
Fix the unit test naming according to the naming convention

0nko committed:
Optimize imports
